### ISSUE TITLE
Possibly unintentional recursion in `Dry::Schema::Result#message_set`

### ISSUE THREAD

**Issue by aardvark179** [dry-rb/dry-schema#130]
**Describe the bug**
`Dry::Schema::Result#message_set` can call `Dry::Schema::Result#errors`, which recursively calls `message_set`, leading to a stack overflow on TruffleRuby.

**To Reproduce**

On TruffleRuby.

If we create a simple schema such as
```
s = Dry::Schema.define do
  required(:foo) { filled? }
end
```

and execute it with an object on which it will fail like this
```
r = s.call( { } )
```
Then you will get a stack overflow when trying to print the result object.

**Analysis**
`Dry::Schema::Result#message_set` calls the message compiler with the result ast. If the ast is not empty then the message compiler will do
```
        current_messages = EMPTY_ARRAY.dup
        compiled_messages = ast.map { |node| visit(node, EMPTY_OPTS.dup(current_messages)) }
```
The `visit` call resolves to a `Dry::Schema::MessageCompiler.visit_predicate` call, which builds an `options` hash containing various bits of information including `:input` which points to the original result object. It then does `message[predicate, options]` which results to `Dry::Schema::Messages::Abstract#call` which does `cache.fetch_or_store(args.hash) do ... end`. That attempt to calculate a hash of the splatted arguments will recursively `hash` on the `options` passed in ultimately hitting the original result object, whose hash method is defined vai `Dry::Equalizer` to be based on `output` and `errors`. Calling errors calls `message_set`, and we recurse round the circle again.

The stack overflow does not occur on MRI because `Array#hash` and `Hash#hash` use `rb_hash` internally and that checks exhaustively for recursion and substitutes `0` for recursive hashes.

We will look at fixing the recursion guards in TruffleRuby, but that is probably simply hiding an underlying issue as I doubt `message_set` and `errors` were intended to recursively call each other.

**Expected behavior**

`message_set` and `errors` should not recursively call each other.

**Your environment**

- Affects my production application: **NO**
- Ruby version: TruffleRuby RC17
- OS: Oracle Linux 7.5

**Comment by solnic** [comment]
Thanks for this detailed investigation. I'll fix this in 1.0.1 soon.

**Comment by solnic** [comment]
@aardvark179 I've fixed it but it seems like travis is unable to install truffleruby https://travis-ci.org/dry-rb/dry-schema/jobs/529697296#L483
